docs/sphinx/hpx-functions.rst

```diff
@@ -0,0 +1,102 @@
+.. _HPX Functions:
+
+=============
+HPX Functions
+=============
+
+This document describes the prototypes, the allowable arguments of the listed
+HPX functions.
+
+async
+=====
+
+Async in C++ allows running asynchronous programming. Async programming is useful
+in compute heavy implementations. Basic async function would look like
+
+.. code-block:: text
+   #include <future>
+   using namespace hpx;
+   int main() {
+    future<int> f = async([]() {});
+    int val = f.get();
+   }
+
+You can chain multiple async operations. For implementing chain async programming
+``objFuture.then(...).then(...).then(...).then(...).then(...)``.
+
+Implementing async functions should have input parameters as
+
+- Lambda function
+- Executor function
+- Launch policy
+
+then
+====
+
+then in asychronous programming is used in chained async implementations. Using
+then function will avoid blocking waits or wasting threads on polling. Future
+can have then chained implementations. With then the antecedent future is ready
+(has a value or exception stored) before the continuation starts as instructed
+by lambda function. Example implementation using hpx is
+
+.. code-block:: text
+
+   #include <future>
+
+   int main() {
+     future<int> fA = async([])(){});
+     future<string> fB = fA.then([](future<int> f)) {
+       return g.get().to_string();
+     });
+   }
+
+
+wait
+====
+
+``hpx::future<T>::wait`` blocks the function implementation until the call is
+available. This function call can be checked using ``valid() == true``. Example
+implementation using hpx is
+
+.. code-block:: text
+
+   #include <future>
+   #include <thread>
+
+   int main(){
+     hpx::future<int> fA = hpx::async([])(){});{
+       func(valA);
+     });
+     hpx::future<string> fB = hpx::async(hpx::launch::async, [](){
+       func(valB);
+     });
+
+     fA.wait();
+     fB.wait();
+
+     hpx::cout << fA.get() << `\n`;
+     hpx::cout << fB.get() << `\n`;
+   }
+
+wait_all
+========
```

Comment:
You can find some documentation of wait_all here: https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/lcos/wait_all.hpp#L18-L100.

Reply:
2c795c4. I still would defer adding api like docs, as those are Doxygen docs/api docs. These docs should be on the line how you would just write as what is ``wait_all`` in hpx like if you would be selling the information or informing the users about the functionality I think. The docs you linked are implementation level docs.

Reply:
Could elaborate on why you prefer not to add API docs?

At the moment:
1. The description you've added is exactly the description from the doxygen docs of `wait_all`. This is bound to get out of sync.
2. I'm not sure where the `literalinclude` is supposed to point, but it doesn't seem to be a reasonable range of lines. If it's supposed to point to the doxygen comments that doesn't make much sense as the text won't be rendered the way it's supposed to (it'll be a *literal* text block).
3. The example is nice, but I think it's incomplete enough to be confusing to users (they can't just copy and paste it and have it work). Best would be to have this as an actual source file in the examples subdirectory (we might have to add a docs/snippets subdirectory or something like that) and `literalinclude` it. This makes sure that the code in the example actually compiles and runs.

I think you could focus on adding doxygen docs for `future` and `async`. Most of the functions already have comments, but they're not in doxygen style (and I think some of them are not quite complete; this is where you can use cppreference). Doxygen docs are not perfect for users, but I think they're a good starting point. We can then continue from there with more free-form content (like a section on task-based programming).

Again, I hope I'm not discouraging you with these comments. I would really recommend you hop onto our IRC channel where we can discuss where to go with this in more detail.